File: frontend/services/polymarketSDK.ts
/**
 * Polymarket SDK Wrapper for Frontend
 * 
 * Обёртка над официальным Polymarket CLOB SDK для TypeScript
 * Статус: Prepared for Integration (требуется подключение кошелька)
 * 
 * Установка зависимостей:
 *   npm install @polymarket/clob-client ethers@5
 * 
 * Использование:
 *   import { PolymarketSDK } from './services/polymarketSDK';
 *   
 *   const sdk = new PolymarketSDK();
 *   await sdk.connectWallet();
 *   const markets = await sdk.getMarkets();
 */

// Типы для Polymarket API
export interface Market {
    id: string;
    question: string;
    outcomes: string[];
    outcomePrices: string[];
    volume: number;
    endDate: string;
    active: boolean;
    image?: string;
}

export interface OrderBook {
    bids: { price: number; size: number }[];
    asks: { price: number; size: number }[];
}

export interface Order {
    id: string;
    token_id: string;
    price: number;
    size: number;
    side: 'BUY' | 'SELL';
    status: 'OPEN' | 'FILLED' | 'CANCELLED';
    createdAt: string;
}

export interface Position {
    market: string;
    outcome: string;
    size: number;
    avgPrice: number;
    pnl: number;
}

export interface Candle {
    timestamp: number;
    open: number;
    high: number;
    low: number;
    close: number;
    volume: number;
}

export type OrderType = 'GTC' | 'GTD' | 'FOK';
export type Resolution = 'minute' | 'hour' | 'day' | 'week';

/**
 * Polymarket SDK Error
 */
export class PolymarketSDKError extends Error {
    constructor(message: string) {
        super(message);
        this.name = 'PolymarketSDKError';
    }
}

/**
 * Polymarket Not Configured Error
 */
export class PolymarketNotConfiguredError extends PolymarketSDKError {
    constructor(message: string) {
        super(message);
        this.name = 'PolymarketNotConfiguredError';
    }
}

/**
 * Polymarket SDK Class
 * 
 * Wrapper над официальным Polymarket CLOB SDK
 * Предоставляет методы для работы с рынками, ордерами и позициями
 */
export class PolymarketSDK {
    private client: any = null;
    private configured: boolean = false;
    private host: string;
    private chainId: number;
    private signer: any = null;
    private apiCreds: any = null;

    constructor(host: string = 'https://clob.polymarket.com', chainId: number = 137) {
        this.host = host;
        this.chainId = chainId;
    }

    /**
     * Проверка настроено ли SDK
     */
    isConfigured(): boolean {
        return this.configured;
    }

    /**
     * Подключение кошелька
     * 
     * @param privateKey - Приватный ключ кошелька
     * @param apiCreds - API credentials (опционально)
     */
    async connectWallet(privateKey: string, apiCreds?: { apiKey: string; apiSecret: string; apiPassphrase: string }): Promise<void> {
        try {
            // Проверяем наличие ethers.js
            if (typeof window === 'undefined' || !(window as any).ethers) {
                throw new PolymarketSDKError('ethers.js not available');
            }

            const ethers = (window as any).ethers;
            
            // Создаём signer из приватного ключа
            this.signer = new ethers.Wallet(privateKey);

            // Импортируем CLOB клиент
            const { ClobClient } = await import('@polymarket/clob-client');
            
            // Создаём клиента
            this.client = new ClobClient(
                this.host,
                this.chainId,
                this.signer,
                apiCreds || undefined
            );

            this.apiCreds = apiCreds || null;
            this.configured = true;

            console.log('✅ PolymarketSDK connected for chain:', this.chainId);
        } catch (error) {
            console.error('❌ Failed to connect wallet:', error);
            throw new PolymarketSDKError(`Failed to connect wallet: ${error}`);
        }
    }

    /**
     * Подключение через браузерный кошелёк (MetaMask, etc.)
     */
    async connectBrowserWallet(): Promise<void> {
        try {
            if (typeof window === 'undefined' || !(window as any).ethereum) {
                throw new PolymarketNotConfiguredError('No crypto wallet found');
            }

            const ethers = (window as any).ethers;
            
            // Запрашиваем доступ к кошельку
            await (window as any).ethereum.request({ method: 'eth_requestAccounts' });
            
            // Создаём signer из provider
            const provider = new ethers.BrowserProvider((window as any).ethereum);
            this.signer = await provider.getSigner();

            // Переключаемся на Polygon
            const chainId = await (window as any).ethereum.request({ method: 'eth_chainId' });
            const polygonChainId = '0x89'; // 137 in hex
            
            if (chainId !== polygonChainId) {
                try {
                    await (window as any).ethereum.request({
                        method: 'wallet_switchEthereumChain',
                        params: [{ chainId: polygonChainId }],
                    });
                } catch (switchError) {
                    // Network needs to be added
                    await (window as any).ethereum.request({
                        method: 'wallet_addEthereumChain',
                        params: [{
                            chainId: polygonChainId,
                            chainName: 'Polygon Mainnet',
                            nativeCurrency: { name: 'MATIC', symbol: 'MATIC', decimals: 18 },
                            rpcUrls: ['https://polygon-rpc.com'],
                            blockExplorerUrls: ['https://polygonscan.com'],
                        }],
                    });
                }
            }

            // Импортируем CLOB клиент
            const { ClobClient } = await import('@polymarket/clob-client');
            
            this.client = new ClobClient(
                this.host,
                this.chainId,
                this.signer
            );

            this.configured = true;
            console.log('✅ PolymarketSDK connected via browser wallet');
        } catch (error) {
            console.error('❌ Failed to connect browser wallet:', error);
            throw new PolymarketSDKError(`Failed to connect browser wallet: ${error}`);
        }
    }

    // ==================== MARKET DATA ====================

    /**
     * Получить список рынков
     */
    async getMarkets(limit: number = 100, active: boolean = true): Promise<Market[]> {
        this.ensureConfigured();
        
        try {
            const markets = await this.client.getMarkets();
            
            // Фильтрация
            let filtered = active ? markets.filter((m: any) => m.active) : markets;
            return filtered.slice(0, limit);
        } catch (error) {
            console.error('❌ Error fetching markets:', error);
            throw new PolymarketSDKError(`Failed to fetch markets: ${error}`);
        }
    }

    /**
     * Получить данные конкретного рынка
     */
    async getMarket(marketId: string): Promise<Market | null> {
        this.ensureConfigured();
        
        try {
            return await this.client.getMarket(marketId);
        } catch (error) {
            console.error(`❌ Error fetching market ${marketId}:`, error);
            return null;
        }
    }

    /**
     * Получить стакан заявок
     */
    async getOrderbook(tokenId: string): Promise<OrderBook> {
        this.ensureConfigured();
        
        try {
            return await this.client.getOrderbook(tokenId);
        } catch (error) {
            console.error(`❌ Error fetching orderbook for ${tokenId}:`, error);
            return { bids: [], asks: [] };
        }
    }

    /**
     * Получить текущую цену токена
     */
    async getPrice(tokenId: string): Promise<number | null> {
        this.ensureConfigured();
        
        try {
            return await this.client.getPrice(tokenId);
        } catch (error) {
            console.error(`❌ Error fetching price for ${tokenId}:`, error);
            return null;
        }
    }

    /**
     * Получить историю цен
     */
    async getPricesHistory(
        tokenId: string,
        resolution: Resolution = 'hour',
        limit: number = 168
    ): Promise<Candle[]> {
        this.ensureConfigured();
        
        try {
            return await this.client.getPricesHistory(tokenId, resolution, limit);
        } catch (error) {
            console.error(`❌ Error fetching price history for ${tokenId}:`, error);
            return [];
        }
    }

    // ==================== ORDERS ====================

    /**
     * Разместить ордер
     */
    async placeOrder(
        tokenId: string,
        price: number,
        size: number,
        side: 'BUY' | 'SELL' = 'BUY',
        orderType: OrderType = 'GTC'
    ): Promise<Order | null> {
        this.ensureConfigured();
        
        try {
            const { OrderArgs, OrderType: ClobOrderType, BUY, SELL } = await import('@polymarket/clob-client');
            
            const orderArgs = new OrderArgs({
                price,
                size,
                side: side === 'BUY' ? BUY : SELL,
                token_id: tokenId
            });
            
            const signedOrder = await this.client.createOrder(orderArgs);
            
            const clobOrderType = orderType === 'GTC' ? ClobOrderType.GTC 
                : orderType === 'GTD' ? ClobOrderType.GTD 
                : ClobOrderType.FOK;
            
            const resp = await this.client.postOrder(signedOrder, clobOrderType);
            
            console.log('✅ Order placed:', resp.orderID);
            return resp;
        } catch (error) {
            console.error('❌ Error placing order:', error);
            throw new PolymarketSDKError(`Failed to place order: ${error}`);
        }
    }

    /**
     * Отменить ордер
     */
    async cancelOrder(orderId: string): Promise<boolean> {
        this.ensureConfigured();
        
        try {
            const resp = await this.client.cancelOrder(orderId);
            console.log('✅ Order cancelled:', orderId);
            return resp.success;
        } catch (error) {
            console.error(`❌ Error cancelling order ${orderId}:`, error);
            return false;
        }
    }

    /**
     * Отменить все ордера
     */
    async cancelAllOrders(): Promise<number> {
        this.ensureConfigured();
        
        try {
            const resp = await this.client.cancelAll();
            const cancelled = resp.orderIDs?.length || 0;
            console.log(`✅ Cancelled ${cancelled} orders`);
            return cancelled;
        } catch (error) {
            console.error('❌ Error cancelling all orders:', error);
            return 0;
        }
    }

    /**
     * Получить ордера пользователя
     */
    async getOrders(marketId?: string): Promise<Order[]> {
        this.ensureConfigured();
        
        try {
            let orders = await this.client.getOrders();
            
            if (marketId) {
                orders = orders.filter((o: any) => o.market === marketId);
            }
            
            return orders;
        } catch (error) {
            console.error('❌ Error fetching orders:', error);
            return [];
        }
    }

    // ==================== POSITIONS ====================

    /**
     * Получить текущие позиции
     */
    async getPositions(): Promise<Position[]> {
        this.ensureConfigured();
        
        try {
            return await this.client.getPositions();
        } catch (error) {
            console.error('❌ Error fetching positions:', error);
            return [];
        }
    }

    /**
     * Получить баланс
     */
    async getBalance(tokenId?: string): Promise<Record<string, number>> {
        this.ensureConfigured();
        
        try {
            const balances = await this.client.getBalances();
            
            if (tokenId) {
                return { [tokenId]: balances[tokenId] || 0 };
            }
            
            return balances;
        } catch (error) {
            console.error('❌ Error fetching balances:', error);
            return {};
        }
    }

    // ==================== UTILITY ====================

    /**
     * Получить время сервера
     */
    async getServerTime(): Promise<Date | null> {
        try {
            return this.client ? await this.client.getTime() : null;
        } catch (error) {
            console.error('❌ Error fetching server time:', error);
            return null;
        }
    }

    /**
     * Получить ставку комиссии
     */
    async getFeeRate(): Promise<number> {
        try {
            return this.client ? await this.client.getFeeRate() : 0;
        } catch (error) {
            console.error('❌ Error fetching fee rate:', error);
            return 0;
        }
    }

    /**
     * Получить tick size для токена
     */
    async getTickSize(tokenId: string): Promise<number> {
        try {
            return this.client ? await this.client.getTickSize(tokenId) : 0.01;
        } catch (error) {
            console.error(`❌ Error fetching tick size for ${tokenId}:`, error);
            return 0.01;
        }
    }

    // ==================== BUILDER SDK (Placeholders) ====================

    /**
     * Подписать транзакцию
     * Требует @polymarket/builder-signing-sdk
     */
    async signTransaction(transaction: any): Promise<string | null> {
        console.warn('⚠️ signTransaction() requires @polymarket/builder-signing-sdk (not installed)');
        return null;
    }

    /**
     * Отправить безгазовую транзакцию
     * Требует @polymarket/builder-relayer-client
     */
    async submitGaslessTransaction(transaction: any, signature: string): Promise<string | null> {
        console.warn('⚠️ submitGaslessTransaction() requires @polymarket/builder-relayer-client (not installed)');
        return null;
    }

    /**
     * Получить объём билдера
     */
    async getBuilderVolume(): Promise<number> {
        console.warn('⚠️ getBuilderVolume() requires Builder SDK (not installed)');
        return 0;
    }

    // ==================== HELPERS ====================

    private ensureConfigured(): void {
        if (!this.configured) {
            throw new PolymarketNotConfiguredError(
                'Polymarket SDK not configured. Call connectWallet() first.'
            );
        }
    }
}

// ==================== GLOBAL INSTANCE ====================

let globalSDK: PolymarketSDK | null = null;

/**
 * Получить глобальный экземпляр SDK
 */
export function getPolymarketSDK(): PolymarketSDK {
    if (globalSDK === null) {
        globalSDK = new PolymarketSDK();
    }
    return globalSDK;
}

/**
 * Проверить настроен ли SDK
 */
export function isPolymarketConfigured(): boolean {
    return globalSDK?.isConfigured() || false;
}

File: api/services/polymarket_sdk.py
"""
Polymarket SDK Wrapper Module

Этот модуль предоставляет обёртку над официальным Python SDK Polymarket CLOB.
Статус: Prepared for Integration (требуется API key для активации)

Установка зависимостей:
    pip install py-clob-client

Требуемые переменные окружения:
    POLYMARKET_API_KEY - API ключ от Polymarket
    POLYMARKET_SECRET - API секрет
    POLYMARKET_PASSPHRASE - API passphrase
    POLYMARKET_PRIVATE_KEY - Приватный ключ кошелька (для подписи)
    POLYMARKET_CHAIN_ID - ID сети (137 для Polygon mainnet)

Пример использования:
    from api.services.polymarket_sdk import PolymarketSDK
    
    sdk = PolymarketSDK()
    if sdk.is_configured():
        markets = sdk.get_markets()
        order = sdk.place_order(token_id="123", price=0.5, size=10, side="BUY")
"""

import os
import logging
from typing import Optional, List, Dict, Any
from datetime import datetime

logger = logging.getLogger(__name__)

# Попытка импорта py-clob-client
try:
    from py_clob_client.client import ClobClient
    from py_clob_client.clob_types import OrderArgs, OrderType, ApiCreds
    from py_clob_client.order_builder.constants import BUY, SELL
    PY_CLOB_AVAILABLE = True
    logger.info("✅ py-clob-client imported successfully")
except ImportError:
    PY_CLOB_AVAILABLE = False
    logger.warning("⚠️ py-clob-client not installed. Run: pip install py-clob-client")


class PolymarketSDKError(Exception):
    """Базовое исключение для Polymarket SDK"""
    pass


class PolymarketNotConfiguredError(PolymarketSDKError):
    """Исключение когда SDK не настроен (нет API ключей)"""
    pass


class PolymarketSDK:
    """
    Wrapper над официальным Polymarket CLOB SDK
    
    Предоставляет методы для:
    - Получения данных рынков (markets, events, prices)
    - Размещения и отмены ордеров
    - Получения информации о позициях пользователя
    - Подписи транзакций (через Builder SDK)
    - Безгазовых транзакций (через Relayer SDK)
    
    Статус: Prepared for Integration
    Для активации необходимо настроить переменные окружения
    """
    
    def __init__(self, host: str = "https://clob.polymarket.com"):
        """
        Инициализация SDK
        
        Args:
            host: URL CLOB API (по умолчанию production)
        """
        self.host = host
        self.client: Optional[ClobClient] = None
        self._configured = False
        
        # Загрузка конфигурации из env
        self.api_key = os.getenv("POLYMARKET_API_KEY")
        self.api_secret = os.getenv("POLYMARKET_SECRET")
        self.api_passphrase = os.getenv("POLYMARKET_PASSPHRASE")
        self.private_key = os.getenv("POLYMARKET_PRIVATE_KEY")
        self.chain_id = int(os.getenv("POLYMARKET_CHAIN_ID", "137"))
        self.signature_type = int(os.getenv("POLYMARKET_SIGNATURE_TYPE", "0"))
        self.funder = os.getenv("POLYMARKET_FUNDER")
        
        # Проверяем конфигурацию
        self._check_configuration()
    
    def _check_configuration(self) -> None:
        """Проверяет наличие необходимых API ключей"""
        has_keys = all([
            self.api_key,
            self.api_secret,
            self.api_passphrase,
            self.private_key
        ])
        
        if has_keys and PY_CLOB_AVAILABLE:
            self._configured = True
            logger.info("✅ PolymarketSDK configured successfully")
        else:
            self._configured = False
            if not PY_CLOB_AVAILABLE:
                logger.warning("⚠️ py-clob-client package not installed")
            else:
                logger.warning("⚠️ Polymarket API keys not configured")
    
    def is_configured(self) -> bool:
        """
        Проверяет настроено ли SDK
        
        Returns:
            bool: True если SDK готов к работе
        """
        return self._configured
    
    def _ensure_configured(self) -> None:
        """Гарантирует что SDK настроен, иначе выбрасывает исключение"""
        if not self._configured:
            raise PolymarketNotConfiguredError(
                "Polymarket SDK not configured. "
                "Please set POLYMARKET_API_KEY, POLYMARKET_SECRET, "
                "POLYMARKET_PASSPHRASE, and POLYMARKET_PRIVATE_KEY environment variables."
            )
    
    def _get_client(self) -> ClobClient:
        """
        Получает или создаёт CLOB клиента
        
        Returns:
            ClobClient: Настроенный клиент
            
        Raises:
            PolymarketNotConfiguredError: Если SDK не настроен
        """
        self._ensure_configured()
        
        if self.client is None:
            try:
                from py_clob_client.client import ClobClient
                
                self.client = ClobClient(
                    host=self.host,
                    key=self.private_key,
                    chain_id=self.chain_id,
                    signature_type=self.signature_type,
                    funder=self.funder if self.signature_type != 0 else None
                )
                
                # Устанавливаем API credentials
                api_creds = ApiCreds(
                    api_key=self.api_key,
                    api_secret=self.api_secret,
                    api_passphrase=self.api_passphrase
                )
                self.client.set_api_creds(api_creds)
                
                logger.info(f"✅ ClobClient initialized for chain_id={self.chain_id}")
                
            except Exception as e:
                logger.error(f"❌ Failed to initialize ClobClient: {e}")
                raise PolymarketSDKError(f"Failed to initialize client: {e}")
        
        return self.client
    
    # ==================== MARKET DATA ====================
    
    def get_markets(self, limit: int = 100, active: bool = True) -> List[Dict[str, Any]]:
        """
        Получить список рынков
        
        Args:
            limit: Максимальное количество рынков
            active: Только активные рынки
            
        Returns:
            Список словарей с данными рынков
            
        Raises:
            PolymarketNotConfiguredError: Если SDK не настроен
        """
        self._ensure_configured()
        
        try:
            client = self._get_client()
            markets = client.get_markets()
            
            # Фильтрация
            if active:
                markets = [m for m in markets if m.get('active', True)]
            
            return markets[:limit]
            
        except Exception as e:
            logger.error(f"❌ Error fetching markets: {e}")
            raise PolymarketSDKError(f"Failed to fetch markets: {e}")
    
    def get_market(self, market_id: str) -> Optional[Dict[str, Any]]:
        """
        Получить данные конкретного рынка
        
        Args:
            market_id: ID рынка
            
        Returns:
            Данные рынка или None
        """
        self._ensure_configured()
        
        try:
            client = self._get_client()
            return client.get_market(market_id)
        except Exception as e:
            logger.error(f"❌ Error fetching market {market_id}: {e}")
            return None
    
    def get_orderbook(self, token_id: str) -> Dict[str, Any]:
        """
        Получить стакан заявок для токена
        
        Args:
            token_id: ID токена
            
        Returns:
            Стакан с bids и asks
        """
        self._ensure_configured()
        
        try:
            client = self._get_client()
            return client.get_orderbook(token_id)
        except Exception as e:
            logger.error(f"❌ Error fetching orderbook for {token_id}: {e}")
            return {"bids": [], "asks": []}
    
    def get_price(self, token_id: str) -> Optional[float]:
        """
        Получить текущую цену токена
        
        Args:
            token_id: ID токена
            
        Returns:
            Цена токена или None
        """
        self._ensure_configured()
        
        try:
            client = self._get_client()
            return client.get_price(token_id)
        except Exception as e:
            logger.error(f"❌ Error fetching price for {token_id}: {e}")
            return None
    
    def get_prices_history(self, token_id: str, resolution: str = "hour", limit: int = 168) -> List[Dict]:
        """
        Получить историю цен
        
        Args:
            token_id: ID токена
            resolution: Разрешение ('minute', 'hour', 'day', 'week')
            limit: Количество точек данных
            
        Returns:
            Список свечей [timestamp, open, high, low, close, volume]
        """
        self._ensure_configured()
        
        try:
            client = self._get_client()
            return client.get_prices_history(token_id, resolution, limit)
        except Exception as e:
            logger.error(f"❌ Error fetching price history for {token_id}: {e}")
            return []
    
    # ==================== ORDERS ====================
    
    def place_order(
        self,
        token_id: str,
        price: float,
        size: float,
        side: str = "BUY",
        order_type: str = "GTC"
    ) -> Optional[Dict[str, Any]]:
        """
        Разместить ордер
        
        Args:
            token_id: ID токена
            price: Цена ордера (0.0 - 1.0)
            size: Размер ордера (количество токенов)
            side: Сторона ордера ('BUY' или 'SELL')
            order_type: Тип ордера ('GTC', 'GTD', 'FOK')
            
        Returns:
            Данные размещённого ордера или None
            
        Raises:
            PolymarketNotConfiguredError: Если SDK не настроен
        """
        self._ensure_configured()
        
        try:
            client = self._get_client()
            
            # Определяем сторону
            clob_side = BUY if side.upper() == "BUY" else SELL
            
            # Создаём аргументы ордера
            order_args = OrderArgs(
                price=price,
                size=size,
                side=clob_side,
                token_id=token_id
            )
            
            # Подписываем ордер
            signed_order = client.create_order(order_args)
            
            # Определяем тип ордера
            clob_order_type = OrderType.GTC
            if order_type.upper() == "GTD":
                clob_order_type = OrderType.GTD
            elif order_type.upper() == "FOK":
                clob_order_type = OrderType.FOK
            
            # Размещаем ордер
            resp = client.post_order(signed_order, clob_order_type)
            
            logger.info(f"✅ Order placed: {resp.get('orderID', 'unknown')}")
            return resp
            
        except Exception as e:
            logger.error(f"❌ Error placing order: {e}")
            raise PolymarketSDKError(f"Failed to place order: {e}")
    
    def cancel_order(self, order_id: str) -> bool:
        """
        Отменить ордер
        
        Args:
            order_id: ID ордера
            
        Returns:
            True если успешно
        """
        self._ensure_configured()
        
        try:
            client = self._get_client()
            resp = client.cancel_order(order_id)
            logger.info(f"✅ Order cancelled: {order_id}")
            return resp.get('success', False)
        except Exception as e:
            logger.error(f"❌ Error cancelling order {order_id}: {e}")
            return False
    
    def cancel_all_orders(self) -> int:
        """
        Отменить все ордера
        
        Returns:
            Количество отменённых ордеров
        """
        self._ensure_configured()
        
        try:
            client = self._get_client()
            resp = client.cancel_all()
            cancelled = len(resp.get('orderIDs', []))
            logger.info(f"✅ Cancelled {cancelled} orders")
            return cancelled
        except Exception as e:
            logger.error(f"❌ Error cancelling all orders: {e}")
            return 0
    
    def get_orders(self, market_id: Optional[str] = None) -> List[Dict]:
        """
        Получить ордера пользователя
        
        Args:
            market_id: Опционально фильтр по рынку
            
        Returns:
            Список ордеров
        """
        self._ensure_configured()
        
        try:
            client = self._get_client()
            orders = client.get_orders()
            
            if market_id:
                orders = [o for o in orders if o.get('market') == market_id]
            
            return orders
        except Exception as e:
            logger.error(f"❌ Error fetching orders: {e}")
            return []
    
    # ==================== POSITIONS ====================
    
    def get_positions(self) -> List[Dict]:
        """
        Получить текущие позиции пользователя
        
        Returns:
            Список позиций
        """
        self._ensure_configured()
        
        try:
            client = self._get_client()
            return client.get_positions()
        except Exception as e:
            logger.error(f"❌ Error fetching positions: {e}")
            return []
    
    def get_balance(self, token_id: Optional[str] = None) -> Dict[str, float]:
        """
        Получить баланс пользователя
        
        Args:
            token_id: Опционально ID токена
            
        Returns:
            Словарь с балансами {token_id: amount}
        """
        self._ensure_configured()
        
        try:
            client = self._get_client()
            balances = client.get_balances()
            
            if token_id:
                return {token_id: balances.get(token_id, 0.0)}
            
            return balances
        except Exception as e:
            logger.error(f"❌ Error fetching balances: {e}")
            return {}
    
    # ==================== UTILITY ====================
    
    def get_server_time(self) -> Optional[datetime]:
        """
        Получить время сервера
        
        Returns:
            Время сервера или None
        """
        try:
            client = self._get_client() if self._configured else None
            if client:
                return client.get_time()
            return None
        except Exception as e:
            logger.error(f"❌ Error fetching server time: {e}")
            return None
    
    def get_fee_rate(self) -> float:
        """
        Получить текущую ставку комиссии
        
        Returns:
            Ставка комиссии (в bps)
        """
        try:
            client = self._get_client() if self._configured else None
            if client:
                return client.get_fee_rate()
            return 0.0
        except Exception as e:
            logger.error(f"❌ Error fetching fee rate: {e}")
            return 0.0
    
    def get_tick_size(self, token_id: str) -> float:
        """
        Получить размер шага цены для токена
        
        Args:
            token_id: ID токена
            
        Returns:
            Tick size
        """
        try:
            client = self._get_client() if self._configured else None
            if client:
                return client.get_tick_size(token_id)
            return 0.01
        except Exception as e:
            logger.error(f"❌ Error fetching tick size for {token_id}: {e}")
            return 0.01
    
    # ==================== BUILDER SDK (Placeholder) ====================
    
    def sign_transaction(self, transaction: Dict) -> Optional[str]:
        """
        Подписать транзакцию через Builder Signing SDK
        
        Args:
            transaction: Данные транзакции
            
        Returns:
            Подпись транзакции или None
            
        Note:
            Требуется установка py-builder-signing-sdk
        """
        logger.warning("⚠️ sign_transaction() requires py-builder-signing-sdk (not installed)")
        return None
    
    def submit_gasless_transaction(self, transaction: Dict, signature: str) -> Optional[str]:
        """
        Отправить безгазовую транзакцию через Relayer SDK
        
        Args:
            transaction: Данные транзакции
            signature: Подпись транзакции
            
        Returns:
            Хэш транзакции или None
            
        Note:
            Требуется установка py-builder-relayer-client
        """
        logger.warning("⚠️ submit_gasless_transaction() requires py-builder-relayer-client (not installed)")
        return None
    
    def get_builder_volume(self) -> float:
        """
        Получить объём билдера за день
        
        Returns:
            Объём в USDC
        """
        logger.warning("⚠️ get_builder_volume() requires Builder SDK (not installed)")
        return 0.0


# ==================== GLOBAL INSTANCE ====================

# Глобальный экземпляр SDK (ленивая инициализация)
_polymarket_sdk: Optional[PolymarketSDK] = None


def get_polymarket_sdk() -> PolymarketSDK:
    """
    Получить глобальный экземпляр PolymarketSDK
    
    Returns:
        PolymarketSDK: Настроенный экземпляр SDK
    """
    global _polymarket_sdk
    if _polymarket_sdk is None:
        _polymarket_sdk = PolymarketSDK()
    return _polymarket_sdk


def is_polymarket_configured() -> bool:
    """
    Проверить настроен ли Polymarket SDK
    
    Returns:
        bool: True если настроен
    """
    sdk = get_polymarket_sdk()
    return sdk.is_configured()
